File: python/skewered_protocol/event.py
from __future__ import annotations

from .types import DecodeError, Event, EventPacket, MenuKey, Side, Weapon


def _decode_side(byte: int) -> Side:
    if byte == 0x01:
        return Side.LEFT
    elif byte == 0x02:
        return Side.RIGHT
    elif byte == 0x03:
        return Side.BOTH
    else:
        raise DecodeError("invalid event data")


def _encode_side(side: Side) -> int:
    if side == Side.LEFT:
        return 0x01
    elif side == Side.RIGHT:
        return 0x02
    elif side == Side.BOTH:
        return 0x03
    raise DecodeError("invalid side")  # pragma: no cover


# Event ID -> kind string, and the reverse mapping
_EVENT_ID_TO_KIND: dict[int, str] = {
    0x01: "set_weapon",
    0x02: "enter_menu",
    0x03: "menu_key",
    0x04: "sleep_now",
    0x05: "set_remote_addr",
    0x06: "remote_battery_level",
    0x10: "clear_scores",
    0x11: "score_up",
    0x12: "score_down",
    0x13: "cycle_card",
    0x14: "cycle_p_card",
    0x15: "cycle_priority",
    0x20: "clock_reset",
    0x21: "clock_enter_time",
    0x22: "clock_start_stop",
    0x23: "clock_start_break",
    0x24: "clock_adj_sec",
    0x25: "clock_adj_period",
    0x30: "review_timeline_back",
    0x31: "undo",
    0x32: "review_timeline_fwd",
    0x33: "func",
    0x34: "touche_occurred",
}

_KIND_TO_EVENT_ID: dict[str, int] = {v: k for k, v in _EVENT_ID_TO_KIND.items()}

# Events that carry a Side argument
_SIDE_EVENTS = {"score_up", "score_down", "cycle_card", "cycle_p_card"}

# Events with no extra data (extra byte = 0)
_NO_DATA_EVENTS = {
    "enter_menu", "sleep_now", "clear_scores", "cycle_priority",
    "clock_reset", "clock_enter_time", "clock_start_stop", "clock_start_break",
    "review_timeline_back", "undo", "review_timeline_fwd", "func", "touche_occurred",
}


def decode_event_data(data: bytes | bytearray) -> EventPacket:
    """Decodes 3 raw data bytes into an EventPacket."""
    if len(data) != 3:
        raise DecodeError(f"expected 3 data bytes, got {len(data)}")

    event_id = data[0]
    extra = data[1]
    dropped_count = data[2]

    kind = _EVENT_ID_TO_KIND.get(event_id)
    if kind is None:
        raise DecodeError("invalid event ID")

    if kind == "set_weapon":
        if extra == 0x01:
            event_data = Weapon.SABRE
        elif extra == 0x02:
            event_data = Weapon.EPEE
        elif extra == 0x03:
            event_data = Weapon.FOIL
        else:
            raise DecodeError("invalid event data")
        event = Event(kind=kind, data=event_data)
    elif kind == "menu_key":
        _MENU_KEY_MAP = {
            0x00: MenuKey.OTHER,
            0x01: MenuKey.UP,
            0x02: MenuKey.DOWN,
            0x03: MenuKey.LEFT,
            0x04: MenuKey.RIGHT,
            0x05: MenuKey.SELECT,
            0x06: MenuKey.EXIT,
            0x07: MenuKey.FUNC,
        }
        key = _MENU_KEY_MAP.get(extra)
        if key is None:
            raise DecodeError("invalid event data")
        event = Event(kind=kind, data=key)
    elif kind in _SIDE_EVENTS:
        event = Event(kind=kind, data=_decode_side(extra))
    elif kind == "set_remote_addr":
        event = Event(kind=kind, data=extra)
    elif kind == "remote_battery_level":
        event = Event(kind=kind, data=extra)
    elif kind == "clock_adj_sec":
        # Interpret as signed i8
        signed = extra if extra < 128 else extra - 256
        event = Event(kind=kind, data=signed)
    elif kind == "clock_adj_period":
        signed = extra if extra < 128 else extra - 256
        event = Event(kind=kind, data=signed)
    elif kind in _NO_DATA_EVENTS:
        event = Event(kind=kind)
    else:
        raise DecodeError("invalid event ID")  # pragma: no cover

    return EventPacket(event=event, dropped_count=dropped_count)


def encode_event_data(event: Event, dropped_count: int = 0) -> bytes:
    """Encodes an Event and dropped count into 3 raw data bytes."""
    event_id = _KIND_TO_EVENT_ID.get(event.kind)
    if event_id is None:
        raise DecodeError(f"unknown event kind: {event.kind}")

    extra = 0

    if event.kind == "set_weapon":
        assert isinstance(event.data, Weapon)
        extra = {Weapon.SABRE: 0x01, Weapon.EPEE: 0x02, Weapon.FOIL: 0x03}[event.data]
    elif event.kind == "menu_key":
        assert isinstance(event.data, MenuKey)
        extra = {
            MenuKey.OTHER: 0x00, MenuKey.UP: 0x01, MenuKey.DOWN: 0x02,
            MenuKey.LEFT: 0x03, MenuKey.RIGHT: 0x04, MenuKey.SELECT: 0x05,
            MenuKey.EXIT: 0x06, MenuKey.FUNC: 0x07,
        }[event.data]
    elif event.kind in _SIDE_EVENTS:
        assert isinstance(event.data, Side)
        extra = _encode_side(event.data)
    elif event.kind in ("set_remote_addr", "remote_battery_level"):
        assert isinstance(event.data, int)
        extra = event.data
    elif event.kind in ("clock_adj_sec", "clock_adj_period"):
        assert isinstance(event.data, int)
        extra = event.data & 0xFF

    return bytes([event_id, extra, dropped_count])

File: src/event.rs
use crate::error::DecodeError;
use crate::types::*;

/// A flat event enum representing all wire-level events.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Event {
    SetWeapon(Weapon),
    EnterMenu,
    MenuKey(MenuKey),
    SleepNow,
    SetRemoteAddr(u8),
    RemoteBatteryLevel(u8),
    /// Emulates the remote's MODE/SEL button: the receiver shows the current
    /// weapon if it is idle, or advances to the next weapon if the weapon is
    /// already being shown. The show-vs-advance decision is made by the
    /// receiver (it is time-sensitive), so this event carries no argument.
    NextWeapon,
    /// Emulates a keypress of the physical IR remote. The receiver feeds the
    /// key through its normal IR pipeline, so the press behaves exactly like
    /// the paired remote's button — including mode-dependent routing (config
    /// menu navigation, numeric time entry) and guards (operations blocked
    /// while the clock runs).
    RemoteKey(RemoteKey),

    ClearScores,
    ScoreUp(Side),
    ScoreDown(Side),
    CycleCard(Side),
    CyclePCard(Side),
    CyclePriority,

    ClockReset,
    ClockEnterTime,
    ClockStartStop,
    ClockStartBreak,
    ClockAdjSec(i8),
    ClockAdjPeriod(i8),

    ReviewTimelineBack,
    Undo,
    ReviewTimelineFwd,
    Func,
    ToucheOccurred,
}

/// A decoded event packet: the event plus the dropped count.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct EventPacket {
    pub event: Event,
    pub dropped_count: u8,
}

/// Decodes 3 raw data bytes into an `EventPacket`.
pub fn decode_event_data(data: &[u8; 3]) -> Result<EventPacket, DecodeError> {
    let event = match data[0] {
        0x01 => Event::SetWeapon(match data[1] {
            0x01 => Weapon::Sabre,
            0x02 => Weapon::Epee,
            0x03 => Weapon::Foil,
            _ => return Err(DecodeError::InvalidEventData),
        }),
        0x02 => Event::EnterMenu,
        0x03 => Event::MenuKey(match data[1] {
            0x00 => MenuKey::Other,
            0x01 => MenuKey::Up,
            0x02 => MenuKey::Down,
            0x03 => MenuKey::Left,
            0x04 => MenuKey::Right,
            0x05 => MenuKey::Select,
            0x06 => MenuKey::Exit,
            0x07 => MenuKey::Func,
            _ => return Err(DecodeError::InvalidEventData),
        }),
        0x04 => Event::SleepNow,
        0x05 => Event::SetRemoteAddr(data[1]),
        0x06 => Event::RemoteBatteryLevel(data[1]),
        0x07 => Event::NextWeapon,
        0x08 => {
            Event::RemoteKey(RemoteKey::from_code(data[1]).ok_or(DecodeError::InvalidEventData)?)
        }

        0x10 => Event::ClearScores,
        0x11 => Event::ScoreUp(decode_side(data[1])?),
        0x12 => Event::ScoreDown(decode_side(data[1])?),
        0x13 => Event::CycleCard(decode_side(data[1])?),
        0x14 => Event::CyclePCard(decode_side(data[1])?),
        0x15 => Event::CyclePriority,

        0x20 => Event::ClockReset,
        0x21 => Event::ClockEnterTime,
        0x22 => Event::ClockStartStop,
        0x23 => Event::ClockStartBreak,
        0x24 => Event::ClockAdjSec(data[1] as i8),
        0x25 => Event::ClockAdjPeriod(data[1] as i8),

        0x30 => Event::ReviewTimelineBack,
        0x31 => Event::Undo,
        0x32 => Event::ReviewTimelineFwd,
        0x33 => Event::Func,
        0x34 => Event::ToucheOccurred,

        _ => return Err(DecodeError::InvalidEventId),
    };

    Ok(EventPacket {
        event,
        dropped_count: data[2],
    })
}

/// Encodes an event and dropped count into 3 raw data bytes.
pub fn encode_event_data(event: &Event, dropped_count: u8) -> [u8; 3] {
    let mut data = [0u8; 3];
    data[0] = encode_event_id(event);
    data[1] = encode_event_extra(event);
    data[2] = dropped_count;
    data
}

fn encode_event_id(event: &Event) -> u8 {
    match event {
        Event::SetWeapon(_) => 0x01,
        Event::EnterMenu => 0x02,
        Event::MenuKey(_) => 0x03,
        Event::SleepNow => 0x04,
        Event::SetRemoteAddr(_) => 0x05,
        Event::RemoteBatteryLevel(_) => 0x06,
        Event::NextWeapon => 0x07,
        Event::RemoteKey(_) => 0x08,

        Event::ClearScores => 0x10,
        Event::ScoreUp(_) => 0x11,
        Event::ScoreDown(_) => 0x12,
        Event::CycleCard(_) => 0x13,
        Event::CyclePCard(_) => 0x14,
        Event::CyclePriority => 0x15,

        Event::ClockReset => 0x20,
        Event::ClockEnterTime => 0x21,
        Event::ClockStartStop => 0x22,
        Event::ClockStartBreak => 0x23,
        Event::ClockAdjSec(_) => 0x24,
        Event::ClockAdjPeriod(_) => 0x25,

        Event::ReviewTimelineBack => 0x30,
        Event::Undo => 0x31,
        Event::ReviewTimelineFwd => 0x32,
        Event::Func => 0x33,
        Event::ToucheOccurred => 0x34,
    }
}

fn encode_event_extra(event: &Event) -> u8 {
    match event {
        Event::SetWeapon(w) => match w {
            Weapon::Sabre => 0x01,
            Weapon::Epee => 0x02,
            Weapon::Foil => 0x03,
        },
        Event::MenuKey(key) => match key {
            MenuKey::Other => 0x00,
            MenuKey::Up => 0x01,
            MenuKey::Down => 0x02,
            MenuKey::Left => 0x03,
            MenuKey::Right => 0x04,
            MenuKey::Select => 0x05,
            MenuKey::Exit => 0x06,
            MenuKey::Func => 0x07,
        },
        Event::SetRemoteAddr(id) => *id,
        Event::RemoteBatteryLevel(v) => *v,
        Event::RemoteKey(key) => key.code(),
        Event::ScoreUp(side) => encode_side(side),
        Event::ScoreDown(side) => encode_side(side),
        Event::CycleCard(side) => encode_side(side),
        Event::CyclePCard(side) => encode_side(side),
        Event::ClockAdjSec(v) => *v as u8,
        Event::ClockAdjPeriod(v) => *v as u8,
        _ => 0,
    }
}

fn encode_side(side: &Side) -> u8 {
    match side {
        Side::Left => 0x01,
        Side::Right => 0x02,
        Side::Both => 0x03,
    }
}

fn decode_side(byte: u8) -> Result<Side, DecodeError> {
    match byte {
        0x01 => Ok(Side::Left),
        0x02 => Ok(Side::Right),
        0x03 => Ok(Side::Both),
        _ => Err(DecodeError::InvalidEventData),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clock_start_stop_roundtrip() {
        let data = encode_event_data(&Event::ClockStartStop, 13);
        assert_eq!(data, [0x22, 0, 13]);
        let decoded = decode_event_data(&data).unwrap();
        assert_eq!(decoded.event, Event::ClockStartStop);
        assert_eq!(decoded.dropped_count, 13);
    }

    #[test]
    fn clock_adj_sec_positive() {
        let data = encode_event_data(&Event::ClockAdjSec(1), 0);
        assert_eq!(data, [0x24, 1, 0]);
        let decoded = decode_event_data(&data).unwrap();
        assert_eq!(decoded.event, Event::ClockAdjSec(1));
    }

    #[test]
    fn clock_adj_sec_negative() {
        let data = encode_event_data(&Event::ClockAdjSec(-1), 0);
        assert_eq!(data, [0x24, 255, 0]);
        let decoded = decode_event_data(&data).unwrap();
        assert_eq!(decoded.event, Event::ClockAdjSec(-1));
    }

    #[test]
    fn all_events_roundtrip() {
        let events = [
            Event::SetWeapon(Weapon::Sabre),
            Event::SetWeapon(Weapon::Epee),
            Event::SetWeapon(Weapon::Foil),
            Event::EnterMenu,
            Event::MenuKey(MenuKey::Up),
            Event::MenuKey(MenuKey::Down),
            Event::MenuKey(MenuKey::Left),
            Event::MenuKey(MenuKey::Right),
            Event::MenuKey(MenuKey::Select),
            Event::MenuKey(MenuKey::Exit),
            Event::MenuKey(MenuKey::Func),
            Event::MenuKey(MenuKey::Other),
            Event::SleepNow,
            Event::SetRemoteAddr(42),
            Event::RemoteBatteryLevel(85),
            Event::NextWeapon,
            Event::ClearScores,
            Event::ScoreUp(Side::Left),
            Event::ScoreUp(Side::Right),
            Event::ScoreUp(Side::Both),
            Event::ScoreDown(Side::Left),
            Event::CycleCard(Side::Right),
            Event::CyclePCard(Side::Both),
            Event::CyclePriority,
            Event::ClockReset,
            Event::ClockEnterTime,
            Event::ClockStartStop,
            Event::ClockStartBreak,
            Event::ClockAdjSec(5),
            Event::ClockAdjSec(-3),
            Event::ClockAdjPeriod(1),
            Event::ClockAdjPeriod(-1),
            Event::ReviewTimelineBack,
            Event::Undo,
            Event::ReviewTimelineFwd,
            Event::Func,
            Event::ToucheOccurred,
        ];
        for event in events {
            let data = encode_event_data(&event, 0);
            let decoded = decode_event_data(&data).unwrap();
            assert_eq!(decoded.event, event, "roundtrip failed for {:?}", event);
        }
    }

    #[test]
    fn remote_key_roundtrip() {
        for key in RemoteKey::ALL {
            let event = Event::RemoteKey(key);
            let data = encode_event_data(&event, 0);
            assert_eq!(data, [0x08, key.code(), 0]);
            let decoded = decode_event_data(&data).unwrap();
            assert_eq!(decoded.event, event, "roundtrip failed for {:?}", key);
        }
    }

    #[test]
    fn remote_key_rejects_non_button_codes() {
        for code in [0u8, 1, 130, 255] {
            let data = [0x08, code, 0x00];
            assert_eq!(
                decode_event_data(&data),
                Err(DecodeError::InvalidEventData),
                "code {code} should not decode"
            );
        }
    }

    #[test]
    fn invalid_event_id() {
        let data = [0xFF, 0x00, 0x00];
        assert_eq!(decode_event_data(&data), Err(DecodeError::InvalidEventId));
    }

    #[test]
    fn invalid_weapon_in_event() {
        let data = [0x01, 0x00, 0x00]; // SetWeapon with invalid weapon byte
        assert_eq!(decode_event_data(&data), Err(DecodeError::InvalidEventData));
    }

    #[test]
    fn invalid_side_in_event() {
        let data = [0x11, 0x00, 0x00]; // ScoreUp with invalid side byte
        assert_eq!(decode_event_data(&data), Err(DecodeError::InvalidEventData));
    }

    #[test]
    fn dropped_count_capping() {
        // Verify we can encode/decode max dropped count
        let data = encode_event_data(&Event::ClockStartStop, 250);
        let decoded = decode_event_data(&data).unwrap();
        assert_eq!(decoded.dropped_count, 250);
    }
}
